Implement new "default" publisher

davidfowl opened the pull request:
## Description

- This publisher walks the model and tries to publish the resource if there's a specific annotation.
- Added supporting APIs for adding a publisher callback (mimics) the manifest publisher.
- Don't register publishers for k8s and docker compose anymore, just make those environment resources add publishing annotations.
- DockerComposeEnvironmentResource and KuberentesEnvironmentResource can publish
- GetDeploymentTarget can return the right target for the annotation

Contributes to #9089

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [x] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [x] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [x] No
  - [ ] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue (consider using one of the following templates):
      - [New (or update) `doc-idea` template](https://github.com/dotnet/docs-aspire/issues/new?template=02-docs-request.yml)
      - [New `breaking-change` template](https://github.com/dotnet/docs-aspire/issues/new?template=04-breaking-change.yml)
      - [New `diagnostic` template](https://github.com/dotnet/docs-aspire/issues/new?template=06-diagnostic-addition.yml)
  - [x] No

davidfowl committed:
Implement new "default" publisher
- This publisher walks the model and tries to publish the resource if there's a specific annotation.
- Added supporting APIs for adding a publisher callback (mimics) the manifest publisher.
- Don't register publishers for k8s and docker compose anymore, just make those environment resources add publishing annotations.
- DockerComposeEnvironmentResource and KuberentesEnvironmentResource can publish
- GetDeploymentTarget can return the right target for the annotation

davidfowl committed:
Remove redundant pragma warning directives in ResourceExtensions.cs

davidfowl committed:
Fix error in tests

mitchdenny commented on src/Aspire.Hosting/ResourceBuilderExtensions.cs:
```diff
@@ -283,6 +283,22 @@ public static IResourceBuilder<T> WithManifestPublishingCallback<T>(this IResour
+    /// <param name="builder">The resource builder.</param>
+    /// <param name="callback">Callback method which takes a <see cref="DefaultPublishingContext"/> which can be used to publish assets.</param>
+    /// <returns></returns>
+    public static IResourceBuilder<T> WithDefaultPublishingCallback<T>(this IResourceBuilder<T> builder, Func<DefaultPublishingContext, Task> callback) where T : IResource
```
I'm not sure about the name default. I'm wondering whether we should just have PublishingContext, WithPublishingCallback etc.

davidfowl replied:
Done

mitchdenny reviewed (approved):
I like where this is going. Lets get it merged.

davidfowl committed:
Refactor publishing annotations and context for improved clarity and functionality

davidfowl committed:
Reorganize using directives and namespace declaration for clarity

davidfowl committed:
Reorganize using directives for improved clarity

davidfowl committed:
Enhance GetDeploymentTargetAnnotation to prioritize ComputeEnvironmentAnnotation when available

davidfowl committed:
Refactor WriteKubernetesOutputAsync to improve handling of Kubernetes environments and enhance error messaging

davidfowl committed:
Fix tests

eerhardt commented on src/Aspire.Hosting/ApplicationModel/ResourceExtensions.cs:
```diff
@@ -575,35 +575,34 @@ public static int GetReplicaCount(this IResource resource)
-                .OfType<DeploymentTargetAnnotation>()
-                .LastOrDefault(a => a.ComputeEnvironment == computeEnvironmentAnnotation.ComputeEnvironment);
+            // If the resource has a ComputeEnvironmentAnnotation, use it to get the compute environment.
+            // This wins over the specified computeEnvironmentResource
```
This seems like odd behavior. I explicitly said "environmentFoo", but the method return "environmentBar".

davidfowl replied:
Now the environmentResource is a filter.

eerhardt commented on src/Aspire.Hosting/ApplicationModel/PublishingCallbackAnnotation.cs:
```diff
@@ -0,0 +1,19 @@
+/// Initializes a new instance of the <see cref="PublishingCallbackAnnotation"/> class.
+/// </remarks>
+/// <param name="callback">The publishing callback.</param>
+public class PublishingCallbackAnnotation(Func<PublishingContext, Task> callback) : IResourceAnnotation
```
sealed?

davidfowl commented on playground/publishers/Publishers.AppHost/.aspire/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "appHostPath": "../Publishers.AppHost.csproj"
```
CLI dumps this turd to stop scanning for the app host on every run.

eerhardt replied:
That seems .... ick.

davidfowl committed:
Only single compute enviroment at a time
- Change the sample to use a single compute environment switched on a config entry

davidfowl committed:
Make PublishingCallbackAnnotation class sealed to prevent inheritance

davidfowl committed:
Added smoke test for default publisher

davidfowl committed:
Verify that these resources are publishing in the right mode.
